// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/map/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = exports.displayName = exports.name = void 0;
const common_1 = require("../common");
exports.name = 'map';
exports.displayName = 'Map a website to discover all URLs';
exports.operationName = 'map';
exports.resourceName = 'MapSearch';
function createSitemapProperty() {
    return {
        displayName: 'Sitemap',
        name: 'sitemap',
        type: 'options',
        options: [
            {
                name: 'Include',
                value: 'include',
                description: 'Use sitemap plus link discovery (default, most comprehensive)',
            },
            {
                name: 'Only',
                value: 'only',
                description: 'Only return URLs from sitemap.xml (fastest, but may miss pages)',
            },
            {
                name: 'Skip',
                value: 'skip',
                description: 'Ignore sitemap, discover pages through links only',
            },
        ],
        default: 'include',
        description: 'Control how URLs are discovered. "Include" combines sitemap with link crawling for best coverage. "Only" is fastest but limited to sitemap. "Skip" relies purely on link following.',
        routing: {
            request: {
                body: {
                    sitemap: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createIncludeSubdomainsProperty() {
    return {
        displayName: 'Include Subdomains',
        name: 'includeSubdomains',
        type: 'boolean',
        default: false,
        description: 'Whether to discover URLs on subdomains like blog.example.com or docs.example.com. Enable to map the entire domain ecosystem, not just the main domain.',
        routing: {
            request: {
                body: {
                    includeSubdomains: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createLimitProperty() {
    return {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        typeOptions: {
            minValue: 1,
            maxValue: 5000,
        },
        default: 5000,
        description: 'Max number of results to return',
        routing: {
            request: {
                body: {
                    limit: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createTimeoutProperty() {
    return {
        displayName: 'Timeout (Ms)',
        name: 'timeout',
        type: 'number',
        default: 10000,
        description: 'Maximum time in milliseconds to wait for URL discovery. Increase for large sites or slow connections. Default 10000ms (10 seconds).',
        routing: {
            request: {
                body: {
                    timeout: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createAdditionalFieldsProperty(operation) {
    return {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        description: 'Additional fields to send in the request body',
        options: [
            {
                displayName: 'Custom Properties (JSON)',
                name: 'customProperties',
                type: 'json',
                default: '{}',
                description: 'Custom JSON properties to add to the request body',
            },
        ],
        routing: {
            request: {
                body: {
                    additionalFields: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        const additionalFields = body.additionalFields;
                        if (additionalFields) {
                            if (additionalFields.customProperties) {
                                try {
                                    const customProps = JSON.parse(additionalFields.customProperties);
                                    Object.assign(requestOptions.body, customProps);
                                }
                                catch (error) {
                                }
                            }
                            delete body.additionalFields;
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            show: {
                operation: [operation],
                useCustomBody: [true],
            },
        },
    };
}
function createMapProperties() {
    return [
        (0, common_1.createOperationNotice)(exports.resourceName, exports.name),
        (0, common_1.createUrlProperty)(exports.name, undefined, exports.resourceName),
        createSitemapProperty(),
        createIncludeSubdomainsProperty(),
        createLimitProperty(),
        createTimeoutProperty(),
    ];
}
const { options, properties } = (0, common_1.buildApiProperties)(exports.name, exports.displayName, createMapProperties());
exports.options = options;
exports.properties = properties;
properties.push(createAdditionalFieldsProperty(exports.name));
