// n8n-nodes-late/dist/late/resources/facebook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.facebookResource = void 0;
const commonFields_1 = require("../utils/commonFields");
exports.facebookResource = {
    operations: [
        {
            name: "List Pages",
            value: "listPages",
            action: "List Facebook pages",
            routing: {
                request: {
                    method: "GET",
                    url: "/connect/facebook/select-page",
                    qs: {
                        profileId: "={{ $parameter.profileId }}",
                        tempToken: "={{ $parameter.tempToken }}",
                    },
                },
            },
        },
        {
            name: "Select Page",
            value: "selectPage",
            action: "Select Facebook page",
            routing: {
                request: {
                    method: "POST",
                    url: "/connect/facebook/select-page",
                    body: {
                        profileId: "={{ $parameter.profileId }}",
                        pageId: "={{ $parameter.pageId }}",
                        tempToken: "={{ $parameter.tempToken }}",
                        userProfile: "={{ { id: $parameter.userProfileId || '', name: $parameter.userProfileName || '', email: $parameter.userProfileEmail || '', picture: $parameter.userProfilePicture || '', accessToken: $parameter.userProfileAccessToken || '' } }}",
                    },
                },
            },
        },
        {
            name: "Update Page",
            value: "updatePage",
            action: "Update Facebook page",
            routing: {
                request: {
                    method: "PUT",
                    url: "=/accounts/{{ $parameter.accountId }}/facebook-page",
                    body: {
                        selectedPageId: "={{ $parameter.selectedPageId }}",
                    },
                },
            },
        },
    ],
    fields: [
        // Common fields using reusable builders
        (0, commonFields_1.buildProfileIdField)("facebook", ["listPages", "selectPage"]),
        (0, commonFields_1.buildTempTokenField)("facebook", ["listPages", "selectPage"]),
        (0, commonFields_1.buildSelectorField)("facebook", ["selectPage"], "pageId", "Page ID", "Facebook page ID to select", undefined, true),
        // User profile fields for OAuth
        ...(0, commonFields_1.buildUserProfileFields)("facebook", ["selectPage"], "facebook"),
        // Update page fields
        (0, commonFields_1.buildAccountIdField)("facebook", ["updatePage"], "Account ID", "Account ID to update"),
        (0, commonFields_1.buildSelectorField)("facebook", ["updatePage"], "selectedPageId", "Selected Page ID", "New page ID to select", undefined, true),
    ],
};
