// n8n-nodes-late/dist/credentials/LateNode.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LateNode = void 0;
class LateNode {
    constructor() {
        this.name = "lateApi";
        this.displayName = "LATE API";
        this.documentationUrl = "https://getlate.dev/docs#authentication";
        this.properties = [
            {
                displayName: "API Key",
                name: "apiKey",
                type: "string",
                typeOptions: { password: true },
                default: "",
                description: 'Your LATE API key. Generate one from your LATE dashboard at getlate.dev/dashboard. Use the raw token without "Bearer " prefix.',
            },
        ];
        // Añadir Authorization: Bearer <token> a cada request
        this.authenticate = {
            type: "generic",
            properties: {
                headers: {
                    Authorization: '={{ "Bearer " + $credentials.apiKey }}',
                },
            },
        };
        // Test de credenciales obligatorio
        this.test = {
            request: {
                baseURL: "https://getlate.dev/",
                url: "/api/v1/usage-stats",
                method: "GET",
            },
        };
    }
}
exports.LateNode = LateNode;
