// n8n-nodes-late/dist/late/utils/routingHooks.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.postsCreatePreSend = postsCreatePreSend;
exports.postsUpdatePreSend = postsUpdatePreSend;
/**
 * Utility functions for routing hooks that process expressions
 */
/**
 * Builds the platforms array from selected platforms and accounts
 */
function buildPlatformsArray(executeFunctions, itemIndex) {
    const selectedPlatforms = executeFunctions.getNodeParameter("selectedPlatforms", itemIndex, []);
    return selectedPlatforms
        .map((platform) => {
        const accountsParam = `${platform}Accounts`;
        const accounts = executeFunctions.getNodeParameter(accountsParam, itemIndex, []);
        return accounts.map((id) => ({
            platform,
            accountId: id,
        }));
    })
        .flat();
}
/**
 * Builds the tags array from comma-separated string
 */
function buildTagsArray(executeFunctions, itemIndex, allowUndefined = false) {
    const tags = executeFunctions.getNodeParameter("tags", itemIndex, "");
    if (!tags)
        return allowUndefined ? undefined : [];
    return tags
        .split(",")
        .map((tag) => tag.trim())
        .filter((tag) => tag);
}
/**
 * Builds TikTok settings if TikTok is selected
 */
function buildTikTokSettings(executeFunctions, itemIndex) {
    const selectedPlatforms = executeFunctions.getNodeParameter("selectedPlatforms", itemIndex, []);
    if (!selectedPlatforms.includes("tiktok"))
        return undefined;
    return {
        privacyLevel: executeFunctions.getNodeParameter("tiktokPrivacyLevel", itemIndex, "PUBLIC_TO_EVERYONE"),
        allowComments: executeFunctions.getNodeParameter("tiktokAllowComments", itemIndex, true),
        allowDuet: executeFunctions.getNodeParameter("tiktokAllowDuet", itemIndex, true),
        allowStitch: executeFunctions.getNodeParameter("tiktokAllowStitch", itemIndex, true),
    };
}
/**
 * Pre-send hook for posts create operation
 */
async function postsCreatePreSend(requestOptions) {
    const { processParametersWithExpressions } = await Promise.resolve().then(() => __importStar(require("./expressionProcessor")));
    // Process expressions in parameters that need it
    const processedParams = processParametersWithExpressions("posts", "create", this, 0);
    // Build the body with processed parameters
    requestOptions.body = {
        content: this.getNodeParameter("content", 0),
        platforms: buildPlatformsArray(this, 0),
        scheduledFor: this.getNodeParameter("scheduledFor", 0, undefined),
        timezone: this.getNodeParameter("timezone", 0, "UTC"),
        publishNow: this.getNodeParameter("publishNow", 0, false),
        isDraft: this.getNodeParameter("isDraft", 0, false),
        visibility: this.getNodeParameter("visibility", 0, "public"),
        tags: buildTagsArray(this, 0),
        mediaItems: processedParams.mediaItems?.items || [],
        twitterThread: processedParams.twitterThreadItems?.items || [],
        threadsConversation: processedParams.threadsThreadItems?.items || [],
        tiktokSettings: buildTikTokSettings(this, 0),
    };
    return requestOptions;
}
/**
 * Pre-send hook for posts update operation
 */
async function postsUpdatePreSend(requestOptions) {
    const { processParametersWithExpressions } = await Promise.resolve().then(() => __importStar(require("./expressionProcessor")));
    const processedParams = processParametersWithExpressions("posts", "update", this, 0);
    requestOptions.body = {
        content: this.getNodeParameter("content", 0),
        platforms: buildPlatformsArray(this, 0),
        scheduledFor: this.getNodeParameter("scheduledFor", 0),
        timezone: this.getNodeParameter("timezone", 0),
        publishNow: this.getNodeParameter("publishNow", 0),
        isDraft: this.getNodeParameter("isDraft", 0),
        visibility: this.getNodeParameter("visibility", 0),
        tags: buildTagsArray(this, 0, true), // Allow undefined for updates
        mediaItems: processedParams.mediaItems?.items || [],
        twitterThread: processedParams.twitterThreadItems?.items || [],
        threadsConversation: processedParams.threadsThreadItems?.items || [],
        tiktokSettings: buildTikTokSettings(this, 0),
    };
    return requestOptions;
}
