// n8n-nodes-late/dist/late/Late.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Late = void 0;
const nodeBuilder_1 = require("./utils/nodeBuilder");
const platformHelpers_1 = require("./utils/platformHelpers");
const resources_1 = require("./resources");
class Late {
    constructor() {
        this.methods = {
            loadOptions: Object.fromEntries(platformHelpers_1.SUPPORTED_PLATFORMS.map((platform) => [
                `get${platform.value.charAt(0).toUpperCase() + platform.value.slice(1)}Accounts`,
                async function () {
                    return (0, platformHelpers_1.loadPlatformAccounts)(this, platform.value);
                },
            ])),
        };
        this.description = {
            displayName: "LATE",
            name: "late",
            icon: "file:late-logo.svg",
            group: ["transform"],
            version: 1,
            description: "Schedule and manage social media posts across multiple platforms with LATE - the professional social media management platform supporting Twitter/X, Instagram, Facebook, LinkedIn, TikTok, YouTube, and Threads",
            subtitle: '={{$parameter["resource"] + ": " + $parameter["operation"]}}',
            defaults: {
                name: "LATE",
            },
            inputs: ["main" /* NodeConnectionType.Main */],
            outputs: ["main" /* NodeConnectionType.Main */],
            credentials: [
                {
                    name: "lateApi",
                    required: true,
                },
            ],
            requestDefaults: {
                baseURL: "https://getlate.dev/api/v1",
                headers: {
                    Accept: "application/json",
                    "Content-Type": "application/json",
                },
            },
            properties: (0, nodeBuilder_1.buildNodeProperties)({
                profiles: resources_1.profilesResource,
                posts: resources_1.postsResource,
                media: resources_1.mediaResource,
                accounts: resources_1.accountsResource,
                connect: resources_1.connectResource,
                usage: resources_1.usageResource,
                facebook: resources_1.facebookResource,
                linkedin: resources_1.linkedinResource,
                clone: resources_1.cloneResource,
            }),
        };
    }
}
exports.Late = Late;
