// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/crawl/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = exports.displayName = exports.name = void 0;
const common_1 = require("../common");
exports.name = 'crawl';
exports.displayName = 'Crawl a website and scrape all pages';
exports.operationName = 'crawl';
exports.resourceName = 'Crawling';
function createExcludePathsProperty(operationName, omitDisplayOptions = false) {
    return {
        displayName: 'Exclude Paths',
        name: 'excludePaths',
        type: 'fixedCollection',
        default: [],
        typeOptions: {
            multipleValues: true,
        },
        description: 'URL path patterns to skip during crawling. Use glob patterns like "blog/*" to exclude sections. Useful for avoiding large archives, user-generated content, or irrelevant pages.',
        placeholder: 'Add path pattern to exclude',
        options: [
            {
                displayName: 'Items',
                name: 'items',
                values: [
                    {
                        displayName: 'Path',
                        name: 'path',
                        type: 'string',
                        default: '',
                        placeholder: 'blog/*',
                        description: 'Path pattern to exclude (e.g., blog/* will exclude paths like /blog/article-1)',
                    },
                ],
            },
        ],
        routing: {
            request: {
                body: {
                    excludePaths: '={{$value.items.map(item => item.path)}}',
                },
            },
        },
        displayOptions: omitDisplayOptions
            ? undefined
            : {
                hide: {
                    useCustomBody: [true],
                },
                show: {
                    resource: [exports.resourceName],
                    operation: [operationName],
                },
            },
    };
}
function createIncludePathsProperty(operationName, omitDisplayOptions = false) {
    return {
        displayName: 'Include Paths',
        name: 'includePaths',
        type: 'fixedCollection',
        default: [],
        typeOptions: {
            multipleValues: true,
        },
        description: 'URL path patterns to include in the crawl. Only pages matching these patterns will be scraped. Use glob patterns like "docs/*" to focus on specific sections.',
        placeholder: 'Add path pattern to include',
        options: [
            {
                displayName: 'Items',
                name: 'items',
                values: [
                    {
                        displayName: 'Path',
                        name: 'path',
                        type: 'string',
                        default: '',
                        placeholder: 'blog/*',
                        description: 'Path pattern to include (e.g., blog/* will only include paths like /blog/article-1)',
                    },
                ],
            },
        ],
        routing: {
            request: {
                body: {
                    includePaths: '={{$value.items.map(item => item.path)}}',
                },
            },
        },
        displayOptions: omitDisplayOptions
            ? undefined
            : {
                hide: {
                    useCustomBody: [true],
                },
                show: {
                    resource: [exports.resourceName],
                    operation: [operationName],
                },
            },
    };
}
function createLimitProperty(operationName) {
    return {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        typeOptions: {
            minValue: 1,
        },
        default: 500,
        description: 'Max number of results to return',
        routing: {
            request: {
                body: {
                    limit: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [operationName],
            },
        },
    };
}
function createDelayProperty(operationName) {
    return {
        displayName: 'Delay',
        name: 'delay',
        type: 'number',
        typeOptions: {
            minValue: 0,
        },
        default: 0,
        description: 'Wait time in milliseconds between scraping each page. Use to be polite to servers or avoid rate limiting. 0 means no delay between requests.',
        routing: {
            request: {
                body: {
                    delay: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        if (body.delay === 0) {
                            delete body.delay;
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [operationName],
            },
        },
    };
}
function createMaxConcurrencyProperty(operationName) {
    return {
        displayName: 'Max Concurrency',
        name: 'maxConcurrency',
        type: 'number',
        typeOptions: {
            minValue: 1,
        },
        default: 100,
        description: 'Maximum number of pages to scrape simultaneously. Higher values complete faster but may trigger rate limits on target sites. Use your team\'s plan limit by default.',
        routing: {
            request: {
                body: {
                    maxConcurrency: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [operationName],
            },
        },
    };
}
function createPromptProperty(operationName) {
    return {
        displayName: 'Prompt',
        name: 'prompt',
        type: 'string',
        default: '',
        description: 'Natural language instructions to guide the crawl. Use to specify what content to focus on, pages to prioritize, or extraction goals (e.g., "Focus on product pages and pricing information").',
        routing: {
            request: {
                body: {
                    prompt: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [operationName],
            },
        },
    };
}
function createCrawlOptionsProperty(operationName) {
    return {
        displayName: 'Crawl Options',
        name: 'crawlOptions',
        type: 'collection',
        placeholder: 'Add Option',
        default: {},
        options: [
            {
                displayName: 'Ignore Sitemap',
                name: 'ignoreSitemap',
                type: 'boolean',
                default: false,
                description: 'Whether to skip reading the website\'s sitemap.xml. Enable if the sitemap is inaccurate, outdated, or you want to discover pages through link following only.',
                routing: {
                    request: {
                        body: {
                            ignoreSitemap: '={{ $value }}',
                        },
                    },
                },
            },
            {
                displayName: 'Ignore Query Params',
                name: 'ignoreQueryParameters',
                type: 'boolean',
                default: false,
                description: 'Whether to treat URLs with different query parameters as the same page. Enable to avoid duplicate scrapes of pages like /products?page=1 and /products?page=2.',
                routing: {
                    request: {
                        body: {
                            ignoreQueryParameters: '={{ $value }}',
                        },
                    },
                },
            },
            {
                displayName: 'Allow External Links',
                name: 'allowExternalLinks',
                type: 'boolean',
                default: false,
                description: 'Whether to follow links to other domains during the crawl. Use with caution as this can significantly expand the crawl scope and credit usage.',
                routing: {
                    request: {
                        body: {
                            allowExternalLinks: '={{ $value }}',
                        },
                    },
                },
            },
            {
                displayName: 'Allow Subdomains',
                name: 'allowSubdomains',
                type: 'boolean',
                default: false,
                description: 'Whether to include subdomains like blog.example.com or docs.example.com when crawling example.com. Enable to capture content across the entire domain ecosystem.',
                routing: {
                    request: {
                        body: {
                            allowSubdomains: '={{ $value }}',
                        },
                    },
                },
            },
        ],
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [operationName],
            },
        },
    };
}
function createAdditionalFieldsProperty(operation) {
    return {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        description: 'Additional fields to send in the request body',
        options: [
            {
                displayName: 'Custom Properties (JSON)',
                name: 'customProperties',
                type: 'json',
                default: '{}',
                description: 'Custom JSON properties to add to the request body',
            },
        ],
        routing: {
            request: {
                body: {
                    additionalFields: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        const additionalFields = body.additionalFields;
                        if (additionalFields) {
                            if (additionalFields.customProperties) {
                                try {
                                    const customProps = JSON.parse(additionalFields.customProperties);
                                    Object.assign(requestOptions.body, customProps);
                                }
                                catch (error) {
                                }
                            }
                            delete body.additionalFields;
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            show: {
                operation: [operation],
                useCustomBody: [true],
            },
        },
    };
}
function createCrawlProperties() {
    return [
        (0, common_1.createOperationNotice)(exports.resourceName, exports.name),
        (0, common_1.createUrlProperty)(exports.name, 'https://firecrawl.dev', exports.resourceName),
        createPromptProperty(exports.operationName),
        createLimitProperty(exports.operationName),
        createDelayProperty(exports.operationName),
        createMaxConcurrencyProperty(exports.operationName),
        createExcludePathsProperty(exports.operationName),
        createIncludePathsProperty(exports.operationName),
        createCrawlOptionsProperty(exports.operationName),
        (0, common_1.createScrapeOptionsProperty)(exports.operationName, true, false, exports.resourceName),
    ];
}
const { options, properties } = (0, common_1.buildApiProperties)(exports.name, exports.displayName, createCrawlProperties());
exports.options = options;
exports.properties = properties;
properties.push(createAdditionalFieldsProperty(exports.name));
