// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/search/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = exports.displayName = exports.name = void 0;
const common_1 = require("../common");
const common_2 = require("../common");
exports.name = 'search';
exports.displayName = 'Search the web and scrape results';
exports.operationName = 'search';
exports.resourceName = 'MapSearch';
function createQueryProperty(operationName) {
    return {
        displayName: 'Query',
        name: 'query',
        type: 'string',
        default: '',
        required: true,
        description: 'The search query to find relevant web pages. Use natural language like you would in a search engine. Results are scraped and returned with their content.',
        routing: {
            request: {
                body: {
                    query: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            show: {
                operation: [operationName],
            },
        },
    };
}
function createLimitProperty(operationName) {
    return {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        typeOptions: {
            minValue: 1,
        },
        default: 5,
        description: 'Max number of results to return',
        routing: {
            request: {
                body: {
                    limit: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            show: {
                operation: [operationName],
            },
            hide: {
                useCustomBody: [true],
            },
        },
    };
}
function createSourcesProperty(operationName, omitDisplayOptions = false) {
    return {
        displayName: 'Sources',
        name: 'sources',
        type: 'multiOptions',
        options: [
            {
                name: 'Web',
                value: 'web',
                description: 'Search general web pages',
            },
            {
                name: 'Images',
                value: 'images',
                description: 'Search for images',
            },
            {
                name: 'News',
                value: 'news',
                description: 'Search news articles',
            },
        ],
        default: ['web'],
        description: 'Types of content to search. Web returns general pages, News focuses on recent articles, Images finds visual content. Select at least one source.',
        routing: {
            request: {
                body: {
                    sources: '={{$value.map(source => ({ type: source }))}}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        if (!body.sources || (Array.isArray(body.sources) && body.sources.length === 0)) {
                            throw new Error('At least one source must be selected (web, images, or news)');
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: omitDisplayOptions
            ? undefined
            : {
                hide: {
                    useCustomBody: [true],
                },
                show: {
                    resource: [exports.resourceName],
                    operation: [operationName],
                },
            },
    };
}
function createTimeBasedSearchProperty(operation) {
    return {
        displayName: 'Time Based Search',
        name: 'tbs',
        type: 'string',
        default: '',
        description: 'Filter results by time. Use "qdr:h" for past hour, "qdr:d" for past day, "qdr:w" for past week, "qdr:m" for past month, "qdr:y" for past year. Leave empty for all time.',
        routing: {
            request: {
                body: {
                    tbs: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            show: {
                operation: [operation],
            },
            hide: {
                useCustomBody: [true],
            },
        },
    };
}
function createAdditionalFieldsProperty(operationName) {
    return {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        description: 'Additional fields to send in the request body',
        options: [
            {
                displayName: 'Custom Properties (JSON)',
                name: 'customProperties',
                type: 'json',
                default: '{}',
                description: 'Custom JSON properties to add to the request body',
            },
        ],
        routing: {
            request: {
                body: {
                    additionalFields: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        const additionalFields = body.additionalFields;
                        if (additionalFields) {
                            if (additionalFields.customProperties) {
                                try {
                                    const customProps = JSON.parse(additionalFields.customProperties);
                                    Object.assign(requestOptions.body, customProps);
                                }
                                catch (error) {
                                }
                            }
                            delete body.additionalFields;
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            show: {
                operation: [operationName],
                useCustomBody: [true],
            },
        },
    };
}
function createTimeoutProperty(operationName) {
    return {
        displayName: 'Timeout (Ms)',
        name: 'timeout',
        type: 'number',
        default: 60000,
        description: 'Maximum time in milliseconds to wait for search results and scraping to complete. Increase for more results or slow connections. Default 60000ms (60 seconds).',
        routing: {
            request: {
                body: {
                    timeout: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [operationName],
            },
        },
    };
}
function createSearchProperties() {
    return [
        (0, common_2.createOperationNotice)(exports.resourceName, exports.name),
        createQueryProperty(exports.name),
        createSourcesProperty(exports.name),
        createTimeoutProperty(exports.name),
        createLimitProperty(exports.name),
        createTimeBasedSearchProperty(exports.name),
    ];
}
const { options, properties } = (0, common_1.buildApiProperties)(exports.name, exports.displayName, createSearchProperties());
exports.options = options;
exports.properties = properties;
properties.push(createAdditionalFieldsProperty(exports.name));
