// n8n-nodes-postiz/dist/credentials/PostizApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostizApi = void 0;
class PostizApi {
    constructor() {
        this.name = 'postizApi';
        this.displayName = 'Postiz API';
        this.documentationUrl = 'https://docs.postiz.com/public-api';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'Host',
                name: 'host',
                type: 'string',
                default: 'https://api.postiz.com',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '={{$credentials.apiKey}}',
                }
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.host}}',
                url: '=/public/v1/is-connected',
            },
        };
    }
}
exports.PostizApi = PostizApi;
