// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiMethods = exports.apiProperties = void 0;
const search_1 = require("./search");
const map_1 = require("./map");
const scrape_1 = require("./scrape");
const crawl_1 = require("./crawl");
const batchScrape_1 = require("./batchScrape");
const batchScrapeStatus_1 = require("./batchScrapeStatus");
const batchScrapeErrors_1 = require("./batchScrapeErrors");
const crawlActive_1 = require("./crawlActive");
const crawlParamsPreview_1 = require("./crawlParamsPreview");
const cancelCrawl_1 = require("./cancelCrawl");
const getCrawlErrors_1 = require("./getCrawlErrors");
const getCrawlStatus_1 = require("./getCrawlStatus");
const extract_1 = require("./extract");
const getExtractStatus_1 = require("./getExtractStatus");
const agent_1 = require("./agent");
const agentAsync_1 = require("./agentAsync");
const getAgentStatus_1 = require("./getAgentStatus");
const teamTokenUsage_1 = require("./teamTokenUsage");
const creditUsageHistorical_1 = require("./creditUsageHistorical");
const cancelBatchScrape_1 = require("./cancelBatchScrape");
const teamCreditUsage_1 = require("./teamCreditUsage");
const teamTokenUsageHistorical_1 = require("./teamTokenUsageHistorical");
const teamQueueStatus_1 = require("./teamQueueStatus");
const scrapingOperationOptions = [
    scrape_1.options,
    batchScrape_1.options,
    batchScrapeStatus_1.options,
    batchScrapeErrors_1.options,
    cancelBatchScrape_1.options,
];
const crawlingOperationOptions = [
    crawl_1.options,
    getCrawlStatus_1.options,
    cancelCrawl_1.options,
    getCrawlErrors_1.options,
    crawlActive_1.options,
    crawlParamsPreview_1.options,
];
const agentOperationOptions = [
    agent_1.options,
    agentAsync_1.options,
    getAgentStatus_1.options,
];
const mapSearchOperationOptions = [
    map_1.options,
    search_1.options,
];
const accountOperationOptions = [
    teamCreditUsage_1.options,
    creditUsageHistorical_1.options,
    teamTokenUsage_1.options,
    teamTokenUsageHistorical_1.options,
    teamQueueStatus_1.options,
];
const extractOperationOptions = [
    extract_1.options,
    getExtractStatus_1.options,
];
const scrapingOperationSelector = {
    displayName: 'Operation',
    name: 'operation',
    type: 'options',
    noDataExpression: true,
    displayOptions: {
        show: {
            resource: ['Scraping'],
        },
    },
    default: 'scrape',
    options: scrapingOperationOptions,
};
const crawlingOperationSelector = {
    displayName: 'Operation',
    name: 'operation',
    type: 'options',
    noDataExpression: true,
    displayOptions: {
        show: {
            resource: ['Crawling'],
        },
    },
    default: 'crawl',
    options: crawlingOperationOptions,
};
const agentOperationSelector = {
    displayName: 'Operation',
    name: 'operation',
    type: 'options',
    noDataExpression: true,
    displayOptions: {
        show: {
            resource: ['Agent'],
        },
    },
    default: 'agent',
    options: agentOperationOptions,
};
const mapSearchOperationSelector = {
    displayName: 'Operation',
    name: 'operation',
    type: 'options',
    noDataExpression: true,
    displayOptions: {
        show: {
            resource: ['MapSearch'],
        },
    },
    default: 'map',
    options: mapSearchOperationOptions,
};
const accountOperationSelector = {
    displayName: 'Operation',
    name: 'operation',
    type: 'options',
    noDataExpression: true,
    displayOptions: {
        show: {
            resource: ['Account'],
        },
    },
    default: 'teamCreditUsage',
    options: accountOperationOptions,
};
const extractOperationSelector = {
    displayName: 'Operation',
    name: 'operation',
    type: 'options',
    noDataExpression: true,
    displayOptions: {
        show: {
            resource: ['Extract'],
        },
    },
    default: 'extract',
    options: extractOperationOptions,
};
const rawProperties = [
    ...scrape_1.properties,
    ...batchScrape_1.properties,
    ...batchScrapeStatus_1.properties,
    ...batchScrapeErrors_1.properties,
    ...cancelBatchScrape_1.properties,
    ...crawl_1.properties,
    ...getCrawlStatus_1.properties,
    ...cancelCrawl_1.properties,
    ...getCrawlErrors_1.properties,
    ...crawlActive_1.properties,
    ...crawlParamsPreview_1.properties,
    ...agent_1.properties,
    ...agentAsync_1.properties,
    ...getAgentStatus_1.properties,
    ...map_1.properties,
    ...search_1.properties,
    ...teamCreditUsage_1.properties,
    ...creditUsageHistorical_1.properties,
    ...teamTokenUsage_1.properties,
    ...teamTokenUsageHistorical_1.properties,
    ...teamQueueStatus_1.properties,
    ...extract_1.properties,
    ...getExtractStatus_1.properties,
];
exports.apiProperties = [
    scrapingOperationSelector,
    crawlingOperationSelector,
    agentOperationSelector,
    mapSearchOperationSelector,
    accountOperationSelector,
    extractOperationSelector,
    ...rawProperties,
];
exports.apiMethods = {
    Scraping: {
        scrape: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        batchScrape: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        batchScrapeStatus: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        batchScrapeErrors: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        cancelBatchScrape: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
    },
    Crawling: {
        crawl: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        getCrawlStatus: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        cancelCrawl: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        getCrawlErrors: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        crawlActive: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        crawlParamsPreview: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
    },
    Agent: {
        agent: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        agentAsync: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        getAgentStatus: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
    },
    MapSearch: {
        map: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        search: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
    },
    Account: {
        teamCreditUsage: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        creditUsageHistorical: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        teamTokenUsage: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        teamTokenUsageHistorical: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        teamQueueStatus: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
    },
    Extract: {
        extract: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
        getExtractStatus: {
            execute() {
                return this.helpers.httpRequest;
            },
        },
    },
};
