// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/batchScrapeErrors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.displayName = exports.name = void 0;
const common_1 = require("../common");
exports.name = 'batchScrapeErrors';
exports.displayName = 'Get batch scrape errors';
exports.resourceName = 'Scraping';
function createBatchIdProperty() {
    return {
        displayName: 'Batch ID',
        name: 'batchId',
        type: 'string',
        required: true,
        default: '',
        description: 'ID of the batch scrape job',
        routing: {
            request: {
                url: '=/batch/scrape/{{$value}}/errors',
            },
        },
        displayOptions: {
            show: {
                resource: [exports.resourceName],
                operation: [exports.name],
            },
        },
    };
}
function createProperties() {
    return [(0, common_1.createOperationNotice)(exports.resourceName, exports.name, 'GET'), createBatchIdProperty()];
}
const { options, properties } = (0, common_1.buildApiProperties)(exports.name, exports.displayName, createProperties());
exports.options = options;
exports.properties = properties;
options.routing = {
    request: {
        method: 'GET',
    },
};
