// n8n-nodes-late/dist/late/resources/linkedin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.linkedinResource = void 0;
const commonFields_1 = require("../utils/commonFields");
exports.linkedinResource = {
    operations: [
        {
            name: "Update Organization",
            value: "updateOrganization",
            action: "Update LinkedIn organization",
            routing: {
                request: {
                    method: "PUT",
                    url: "=/accounts/{{ $parameter.accountId }}/linkedin-organization",
                    body: {
                        accountType: "={{ $parameter.accountType }}",
                        selectedOrganization: "={{ { id: $parameter.organizationId || '', urn: $parameter.organizationUrn || '', name: $parameter.organizationName || '', logoUrl: $parameter.organizationLogoUrl || '', description: $parameter.organizationDescription || '' } }}",
                    },
                },
            },
        },
    ],
    fields: [
        // Common fields using reusable builders
        (0, commonFields_1.buildAccountIdField)("linkedin", ["updateOrganization"], "Account ID", "LinkedIn account ID"),
        (0, commonFields_1.buildAccountTypeField)("linkedin", ["updateOrganization"]),
        // Organization fields
        ...(0, commonFields_1.buildOrganizationFields)("linkedin", ["updateOrganization"]),
    ],
};
