// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/agentAsync/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = void 0;
const common_1 = require("../common");
const name = 'agentAsync';
const displayName = 'Agent (Async) - Returns job ID for manual polling';
exports.operationName = 'agentAsync';
exports.resourceName = 'Agent';
function createPromptProperty() {
    return {
        displayName: 'Prompt',
        name: 'prompt',
        type: 'string',
        typeOptions: {
            rows: 4,
        },
        required: true,
        default: '',
        description: 'Natural language description of the data you want to extract (max 10,000 characters). Be specific about what data you need. Examples: "Find the founders of Firecrawl", "Extract pricing information from this page", "Compare features between these products".',
        placeholder: 'e.g., Find the founders of Firecrawl and their roles',
        routing: {
            request: {
                body: {
                    prompt: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createSpecifyUrlsProperty() {
    return {
        displayName: 'Specify URLs',
        name: 'specifyUrls',
        type: 'boolean',
        default: false,
        description: 'Whether to specify URLs for the agent to focus on. If disabled, the agent will autonomously search and navigate the web to find your data.',
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createUrlsProperty() {
    return {
        displayName: 'URLs',
        name: 'urls',
        type: 'string',
        typeOptions: {
            rows: 4,
        },
        default: '',
        description: 'URLs to focus the agent on specific pages. Accepts multiple formats: a single URL, multiple URLs separated by commas or newlines, or a JSON array like ["url1", "url2"].',
        placeholder: 'https://example.com/page1\nhttps://example.com/page2',
        routing: {
            request: {
                body: {
                    urls: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        if (body.urls !== undefined) {
                            const rawValue = body.urls;
                            if (!rawValue || (typeof rawValue === 'string' && !rawValue.trim())) {
                                delete body.urls;
                                return requestOptions;
                            }
                            const urlsArray = (0, common_1.extractUrls)(rawValue);
                            const uniqueUrls = [...new Set(urlsArray)].filter(Boolean);
                            if (uniqueUrls.length === 0) {
                                delete body.urls;
                            }
                            else {
                                body.urls = uniqueUrls;
                            }
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
                specifyUrls: [true],
            },
        },
    };
}
function createSchemaTypeProperty() {
    return {
        displayName: 'Schema Type',
        name: 'schemaType',
        type: 'options',
        default: 'none',
        description: 'Optionally define a schema for the extracted data structure',
        options: [
            {
                name: 'None',
                value: 'none',
                description: 'No schema - let the agent determine the output structure',
            },
            {
                name: 'Generate From JSON Example',
                value: 'fromExample',
                description: 'Generate a schema from an example JSON object',
            },
            {
                name: 'Define Using JSON Schema',
                value: 'manual',
                description: 'Define the JSON schema manually',
            },
        ],
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createJsonExampleProperty() {
    return {
        displayName: 'JSON Example',
        name: 'jsonExample',
        type: 'json',
        default: '{\n  "founders": [\n    {\n      "name": "John Doe",\n      "role": "CEO"\n    }\n  ]\n}',
        description: 'Provide an example of the JSON structure you want to extract. The schema will be automatically generated from this example.',
        routing: {
            request: {
                body: {
                    jsonExample: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        if (body.jsonExample !== undefined) {
                            try {
                                let example;
                                if (typeof body.jsonExample === 'string') {
                                    const trimmed = body.jsonExample.trim();
                                    if (!trimmed) {
                                        delete body.jsonExample;
                                        return requestOptions;
                                    }
                                    example = JSON.parse(trimmed);
                                }
                                else {
                                    example = body.jsonExample;
                                }
                                body.schema = (0, common_1.convertToSchema)(example);
                                delete body.jsonExample;
                            }
                            catch (error) {
                                throw new Error('Invalid JSON example. Please provide valid JSON.');
                            }
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
                schemaType: ['fromExample'],
            },
        },
    };
}
function createSchemaProperty() {
    return {
        displayName: 'Schema',
        name: 'schema',
        type: 'json',
        default: '{\n  "type": "object",\n  "properties": {\n    "founders": {\n      "type": "array",\n      "items": {\n        "type": "object",\n        "properties": {\n          "name": { "type": "string" },\n          "role": { "type": "string" }\n        }\n      }\n    }\n  }\n}',
        description: 'JSON Schema defining the structure of extracted data. Use JSON Schema format with type definitions. $refs syntax is currently not supported.',
        routing: {
            request: {
                body: {
                    schema: '={{ JSON.parse($value) }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
                schemaType: ['manual'],
            },
        },
    };
}
function createAgentProperties() {
    return [
        (0, common_1.createOperationNotice)(exports.resourceName, name, 'POST'),
        createPromptProperty(),
        createSpecifyUrlsProperty(),
        createUrlsProperty(),
        createSchemaTypeProperty(),
        createJsonExampleProperty(),
        createSchemaProperty(),
    ];
}
const { options, properties } = (0, common_1.buildApiProperties)(name, displayName, createAgentProperties());
exports.options = options;
exports.properties = properties;
options.routing = {
    request: {
        method: 'POST',
        url: '=/agent',
    },
    output: {
        postReceive: [
            {
                type: 'rootProperty',
                properties: {
                    property: 'body',
                },
            },
        ],
    },
};
