// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/scrape/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = exports.displayName = exports.name = void 0;
const common_1 = require("../common");
exports.name = 'scrape';
exports.displayName = 'Scrape a URL and get content as markdown or other formats';
exports.operationName = 'scrape';
exports.resourceName = 'Scraping';
function createParsersProperty(operationName) {
    return {
        displayName: 'Parsers',
        name: 'parsers',
        type: 'multiOptions',
        options: [
            {
                name: 'PDF',
                value: 'pdf',
                description: 'Extract PDF content as markdown (1 credit per page)',
            },
        ],
        default: [],
        description: 'Enable file parsers for content extraction. Select "PDF" to extract PDF content as markdown text. Leave empty to get PDFs as base64 data.',
        routing: {
            request: {
                body: {
                    parsers: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        if (body.parsers !== undefined) {
                            let rawValue = body.parsers;
                            while (Array.isArray(rawValue) && rawValue.length === 1 && Array.isArray(rawValue[0])) {
                                rawValue = rawValue[0];
                            }
                            const extractStrings = (val) => {
                                if (typeof val === 'string') {
                                    try {
                                        const parsed = JSON.parse(val);
                                        return extractStrings(parsed);
                                    }
                                    catch {
                                        return val.trim() ? [val.trim().toLowerCase()] : [];
                                    }
                                }
                                if (Array.isArray(val)) {
                                    return val.flatMap(extractStrings);
                                }
                                if (typeof val === 'object' && val !== null && 'type' in val) {
                                    const typeVal = val.type;
                                    if (typeof typeVal === 'string') {
                                        return [typeVal.toLowerCase()];
                                    }
                                }
                                return [];
                            };
                            const parsersArray = extractStrings(rawValue);
                            const validParsers = ['pdf'];
                            const uniqueParsers = [...new Set(parsersArray)].filter((p) => validParsers.includes(p));
                            if (uniqueParsers.length === 0) {
                                delete body.parsers;
                            }
                            else {
                                body.parsers = uniqueParsers.map((parser) => ({ type: parser }));
                            }
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [operationName],
            },
        },
    };
}
function createAdditionalFieldsProperty(operation) {
    return {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        description: 'Additional fields to send in the request body',
        options: [
            {
                displayName: 'Custom Properties (JSON)',
                name: 'customProperties',
                type: 'json',
                default: '{}',
                description: 'Custom JSON properties to add to the request body',
            },
        ],
        routing: {
            request: {
                body: {
                    additionalFields: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        const additionalFields = body.additionalFields;
                        if (additionalFields) {
                            if (additionalFields.customProperties) {
                                try {
                                    const customProps = JSON.parse(additionalFields.customProperties);
                                    Object.assign(requestOptions.body, customProps);
                                }
                                catch (error) {
                                }
                            }
                            delete body.additionalFields;
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            show: {
                operation: [operation],
                useCustomBody: [true],
            },
        },
    };
}
function createScrapeProperties() {
    return [
        (0, common_1.createOperationNotice)(exports.resourceName, exports.name),
        (0, common_1.createUrlProperty)(exports.name, 'https://firecrawl.dev', exports.resourceName),
        createParsersProperty(exports.operationName),
        (0, common_1.createScrapeOptionsProperty)(exports.operationName, false, false, exports.resourceName),
    ];
}
const { options, properties } = (0, common_1.buildApiProperties)(exports.name, exports.displayName, createScrapeProperties());
exports.options = options;
exports.properties = properties;
properties.push(createAdditionalFieldsProperty(exports.name));
