// n8n-nodes-late/dist/late/resources/clone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloneResource = void 0;
const commonFields_1 = require("../utils/commonFields");
exports.cloneResource = {
    operations: [
        {
            name: "Clone Connection",
            value: "cloneConnection",
            action: "Clone connection to profile",
            routing: {
                request: {
                    method: "POST",
                    url: "=/profiles/{{ $parameter.profileId }}/clone-connection",
                    body: {
                        sourceAccountId: "={{ $parameter.sourceAccountId }}",
                        targetPageId: "={{ $parameter.targetPageId || undefined }}",
                        targetPageName: "={{ $parameter.targetPageName || undefined }}",
                        targetPageAccessToken: "={{ $parameter.targetPageAccessToken || undefined }}",
                        targetOrganizationId: "={{ $parameter.targetOrganizationId || undefined }}",
                        targetOrganizationUrn: "={{ $parameter.targetOrganizationUrn || undefined }}",
                        targetOrganizationName: "={{ $parameter.targetOrganizationName || undefined }}",
                        targetAccountType: "={{ $parameter.targetAccountType || undefined }}",
                    },
                },
            },
        },
    ],
    fields: [
        // Common fields using reusable builders
        {
            ...(0, commonFields_1.buildProfileIdField)("clone", ["cloneConnection"]),
            description: "Target profile ID",
        },
        {
            ...(0, commonFields_1.buildAccountIdField)("clone", ["cloneConnection"], "Source Account ID", "ID of existing connection to clone"),
            name: "sourceAccountId",
        },
        // Facebook-specific fields
        (0, commonFields_1.buildSelectorField)("clone", ["cloneConnection"], "targetPageId", "Target Page ID", "Facebook page ID to target (for Facebook)"),
        (0, commonFields_1.buildSelectorField)("clone", ["cloneConnection"], "targetPageName", "Target Page Name", "Facebook page name (for Facebook)"),
        (0, commonFields_1.buildSelectorField)("clone", ["cloneConnection"], "targetPageAccessToken", "Target Page Access Token", "Facebook page access token (for Facebook)"),
        // LinkedIn-specific fields
        (0, commonFields_1.buildSelectorField)("clone", ["cloneConnection"], "targetOrganizationId", "Target Organization ID", "LinkedIn organization ID (for LinkedIn)", "123456"),
        (0, commonFields_1.buildSelectorField)("clone", ["cloneConnection"], "targetOrganizationUrn", "Target Organization URN", "LinkedIn organization URN (for LinkedIn)", "urn:li:organization:123456"),
        (0, commonFields_1.buildSelectorField)("clone", ["cloneConnection"], "targetOrganizationName", "Target Organization Name", "LinkedIn organization name (for LinkedIn)", "Company Name"),
        // Account type for LinkedIn
        {
            ...(0, commonFields_1.buildAccountTypeField)("clone", ["cloneConnection"]),
            name: "targetAccountType",
            description: "LinkedIn account type (for LinkedIn)",
        },
    ],
};
