// n8n-nodes-late/dist/late/resources/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloneResource = exports.linkedinResource = exports.facebookResource = exports.usageResource = exports.connectResource = exports.accountsResource = exports.mediaResource = exports.postsResource = exports.profilesResource = void 0;
var profiles_1 = require("./profiles");
Object.defineProperty(exports, "profilesResource", { enumerable: true, get: function () { return profiles_1.profilesResource; } });
var posts_1 = require("./posts");
Object.defineProperty(exports, "postsResource", { enumerable: true, get: function () { return posts_1.postsResource; } });
var media_1 = require("./media");
Object.defineProperty(exports, "mediaResource", { enumerable: true, get: function () { return media_1.mediaResource; } });
var accounts_1 = require("./accounts");
Object.defineProperty(exports, "accountsResource", { enumerable: true, get: function () { return accounts_1.accountsResource; } });
var connect_1 = require("./connect");
Object.defineProperty(exports, "connectResource", { enumerable: true, get: function () { return connect_1.connectResource; } });
var usage_1 = require("./usage");
Object.defineProperty(exports, "usageResource", { enumerable: true, get: function () { return usage_1.usageResource; } });
var facebook_1 = require("./facebook");
Object.defineProperty(exports, "facebookResource", { enumerable: true, get: function () { return facebook_1.facebookResource; } });
var linkedin_1 = require("./linkedin");
Object.defineProperty(exports, "linkedinResource", { enumerable: true, get: function () { return linkedin_1.linkedinResource; } });
var clone_1 = require("./clone");
Object.defineProperty(exports, "cloneResource", { enumerable: true, get: function () { return clone_1.cloneResource; } });
