// n8n-nodes-postiz/dist/nodes/Postiz/Postiz.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Postiz = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const GenericFunctions_1 = require("./GenericFunctions");
class Postiz {
    constructor() {
        this.description = {
            displayName: 'Postiz',
            name: 'postiz',
            icon: 'file:postiz.png',
            group: ['output'],
            version: 1,
            subtitle: '={{$parameter["operation"]}}',
            description: 'Consume Postiz API',
            defaults: {
                name: 'Postiz',
            },
            inputs: ["main"],
            outputs: ["main"],
            credentials: [
                {
                    name: 'postizApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Operation',
                    name: 'operation',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Create Post',
                            value: 'createPost',
                            description: 'Schedule a post to Postiz',
                            action: 'Schedule a post to postiz',
                        },
                        {
                            name: 'Delete Post',
                            value: 'deletePost',
                            description: 'Delete a post by ID',
                            action: 'Delete a post by id',
                        },
                        {
                            name: 'Generate Video',
                            value: 'generateVideo',
                            description: 'Generate videos with AI',
                            action: 'Generate videos with AI',
                        },
                        {
                            name: 'Get Channels',
                            value: 'getIntegrations',
                            description: 'Get a list of connected channels',
                            action: 'Get a list of connected channels',
                        },
                        {
                            name: 'Get Posts',
                            value: 'getPosts',
                            description: 'Get a list of posts',
                            action: 'Get a list of posts',
                        },
                        {
                            name: 'Upload File',
                            value: 'uploadFile',
                            description: 'Upload a file to Postiz',
                            action: 'Upload a file to postiz',
                        },
                        {
                            name: 'Video Function',
                            value: 'videoFunction',
                            description: 'Execute video-related functions like loading voices',
                            action: 'Execute video related functions',
                        },
                    ],
                    default: 'createPost',
                },
                {
                    displayName: 'Video Type',
                    name: 'videoType',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['generateVideo'],
                        },
                    },
                    default: 'image-text-slides',
                    required: true,
                    description: 'Type of video to generate (e.g., image-text-slides, veo3)',
                },
                {
                    displayName: 'Output Format',
                    name: 'output',
                    type: 'options',
                    displayOptions: {
                        show: {
                            operation: ['generateVideo'],
                        },
                    },
                    options: [
                        {
                            name: 'Vertical',
                            value: 'vertical',
                        },
                        {
                            name: 'Horizontal',
                            value: 'horizontal',
                        },
                    ],
                    default: 'vertical',
                    required: true,
                    description: 'Video output format',
                },
                {
                    displayName: 'Custom Parameters',
                    name: 'customParameters',
                    placeholder: 'Add Custom Parameter',
                    type: 'fixedCollection',
                    typeOptions: {
                        multipleValues: true,
                    },
                    displayOptions: {
                        show: {
                            operation: ['generateVideo'],
                        },
                    },
                    default: {},
                    options: [
                        {
                            name: 'parameter',
                            displayName: 'Parameter',
                            values: [
                                {
                                    displayName: 'Key',
                                    name: 'key',
                                    type: 'string',
                                    default: '',
                                    required: true,
                                    description: 'Parameter key (e.g., voice, images)',
                                },
                                {
                                    displayName: 'Value',
                                    name: 'value',
                                    type: 'string',
                                    default: '',
                                    required: true,
                                    description: 'Parameter value',
                                },
                            ],
                        },
                    ],
                    description: 'Custom parameters for video generation (e.g., prompt: "description", voice: voice-ID, images: [{"ID":"...","path":"..."}])',
                },
                {
                    displayName: 'Function Name',
                    name: 'functionName',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['videoFunction'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'Video function to execute (e.g., loadVoices)',
                },
                {
                    displayName: 'Identifier',
                    name: 'identifier',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['videoFunction'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'Identifier for the video function (e.g., image-text-slides)',
                },
                {
                    displayName: 'Additional Parameters',
                    name: 'additionalParameters',
                    placeholder: 'Add Parameter',
                    type: 'fixedCollection',
                    typeOptions: {
                        multipleValues: true,
                    },
                    displayOptions: {
                        show: {
                            operation: ['videoFunction'],
                        },
                    },
                    default: {},
                    options: [
                        {
                            name: 'parameter',
                            displayName: 'Parameter',
                            values: [
                                {
                                    displayName: 'Key',
                                    name: 'key',
                                    type: 'string',
                                    default: '',
                                    required: true,
                                    description: 'Parameter key',
                                },
                                {
                                    displayName: 'Value',
                                    name: 'value',
                                    type: 'string',
                                    default: '',
                                    required: true,
                                    description: 'Parameter value',
                                },
                            ],
                        },
                    ],
                    description: 'Additional parameters for the video function',
                },
                {
                    displayName: 'Type',
                    name: 'type',
                    type: 'options',
                    displayOptions: {
                        show: {
                            operation: ['createPost'],
                        },
                    },
                    options: [
                        {
                            name: 'Draft',
                            value: 'draft',
                        },
                        {
                            name: 'Schedule',
                            value: 'schedule',
                        },
                        {
                            name: 'Now',
                            value: 'now',
                        },
                    ],
                    default: 'now',
                    required: true,
                    description: 'Type of post to create',
                },
                {
                    displayName: 'Short Link',
                    name: 'shortLink',
                    type: 'boolean',
                    displayOptions: {
                        show: {
                            operation: ['createPost'],
                        },
                    },
                    default: false,
                    required: true,
                    description: 'Whether to use short links',
                },
                {
                    displayName: 'Date',
                    name: 'date',
                    type: 'dateTime',
                    displayOptions: {
                        show: {
                            operation: ['createPost'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'Date and time for the post',
                },
                {
                    displayName: 'Tags',
                    name: 'tags',
                    placeholder: 'Add Tag',
                    type: 'fixedCollection',
                    typeOptions: {
                        multipleValues: true,
                    },
                    displayOptions: {
                        show: {
                            operation: ['createPost'],
                        },
                    },
                    default: {},
                    options: [
                        {
                            name: 'tag',
                            displayName: 'Tag',
                            values: [
                                {
                                    displayName: 'Value',
                                    name: 'value',
                                    type: 'string',
                                    default: '',
                                    required: true,
                                    description: 'Tag value',
                                },
                                {
                                    displayName: 'Label',
                                    name: 'label',
                                    type: 'string',
                                    default: '',
                                    required: true,
                                    description: 'Tag label',
                                },
                            ],
                        },
                    ],
                    description: 'Tags for the post',
                },
                {
                    displayName: 'Posts',
                    name: 'posts',
                    placeholder: 'Add Post',
                    type: 'fixedCollection',
                    typeOptions: {
                        multipleValues: true,
                    },
                    displayOptions: {
                        show: {
                            operation: ['createPost'],
                        },
                    },
                    default: {},
                    options: [
                        {
                            name: 'post',
                            displayName: 'Post',
                            values: [
                                {
                                    displayName: 'Channel ID',
                                    name: 'integrationId',
                                    type: 'string',
                                    default: '',
                                    required: true,
                                    description: 'ID of the channel (you can get from the get channels operation or from the Postiz UI)',
                                },
                                {
                                    displayName: 'Group',
                                    name: 'group',
                                    type: 'string',
                                    default: '',
                                    description: 'Post group',
                                },
                                {
                                    displayName: 'Settings',
                                    name: 'settings',
                                    placeholder: 'Add Setting',
                                    type: 'fixedCollection',
                                    typeOptions: {
                                        multipleValues: true,
                                    },
                                    default: {},
                                    options: [
                                        {
                                            name: 'setting',
                                            displayName: 'Setting',
                                            values: [
                                                {
                                                    displayName: 'Key',
                                                    name: 'key',
                                                    type: 'string',
                                                    default: '',
                                                    required: true,
                                                    description: 'Setting key',
                                                },
                                                {
                                                    displayName: 'Value',
                                                    name: 'stringValue',
                                                    type: 'string',
                                                    displayOptions: {
                                                        show: {
                                                            valueType: ['string'],
                                                        },
                                                    },
                                                    default: '',
                                                    required: true,
                                                    description: 'String value',
                                                },
                                                {
                                                    displayName: 'Value',
                                                    name: 'numberValue',
                                                    type: 'number',
                                                    displayOptions: {
                                                        show: {
                                                            valueType: ['number'],
                                                        },
                                                    },
                                                    default: 0,
                                                    required: true,
                                                    description: 'Number value',
                                                },
                                                {
                                                    displayName: 'Value',
                                                    name: 'booleanValue',
                                                    type: 'boolean',
                                                    displayOptions: {
                                                        show: {
                                                            valueType: ['boolean'],
                                                        },
                                                    },
                                                    default: false,
                                                    required: true,
                                                    description: 'Whether the setting is enabled',
                                                },
                                                {
                                                    displayName: 'Value',
                                                    name: 'jsonValue',
                                                    type: 'json',
                                                    displayOptions: {
                                                        show: {
                                                            valueType: ['json'],
                                                        },
                                                    },
                                                    default: '',
                                                    required: true,
                                                    description: 'JSON value (object or array)',
                                                },
                                                {
                                                    displayName: 'Value Type',
                                                    name: 'valueType',
                                                    type: 'options',
                                                    options: [
                                                        {
                                                            name: 'String',
                                                            value: 'string',
                                                        },
                                                        {
                                                            name: 'Number',
                                                            value: 'number',
                                                        },
                                                        {
                                                            name: 'Boolean',
                                                            value: 'boolean',
                                                        },
                                                        {
                                                            name: 'JSON',
                                                            value: 'json',
                                                        },
                                                    ],
                                                    default: 'string',
                                                    required: true,
                                                    description: 'Type of the setting value',
                                                },
                                            ],
                                        },
                                    ],
                                    description: 'Provider-specific settings',
                                },
                                {
                                    displayName: 'Content Items',
                                    name: 'value',
                                    placeholder: 'Add Content Item',
                                    type: 'fixedCollection',
                                    typeOptions: {
                                        multipleValues: true,
                                    },
                                    default: {},
                                    options: [
                                        {
                                            name: 'contentItem',
                                            displayName: 'Content Item',
                                            values: [
                                                {
                                                    displayName: 'Content',
                                                    name: 'content',
                                                    type: 'string',
                                                    typeOptions: {
                                                        rows: 4,
                                                    },
                                                    default: '',
                                                    required: true,
                                                    description: 'Content text',
                                                },
                                                {
                                                    displayName: 'ID',
                                                    name: 'id',
                                                    type: 'string',
                                                    default: '',
                                                    description: 'Content ID (optional)',
                                                },
                                                {
                                                    displayName: 'Images',
                                                    name: 'image',
                                                    placeholder: 'Add Image',
                                                    type: 'fixedCollection',
                                                    typeOptions: {
                                                        multipleValues: true,
                                                    },
                                                    default: {},
                                                    options: [
                                                        {
                                                            name: 'imageItem',
                                                            displayName: 'Image',
                                                            values: [
                                                                {
                                                                    displayName: 'ID',
                                                                    name: 'id',
                                                                    type: 'string',
                                                                    default: '',
                                                                    required: true,
                                                                    description: 'Image ID',
                                                                },
                                                                {
                                                                    displayName: 'Path',
                                                                    name: 'path',
                                                                    type: 'string',
                                                                    default: '',
                                                                    required: true,
                                                                    description: 'Image path/URL',
                                                                },
                                                            ],
                                                        },
                                                    ],
                                                    description: 'Images for this content item',
                                                },
                                            ],
                                        },
                                    ],
                                    description: 'Content items (value array)',
                                },
                            ],
                        },
                    ],
                    description: 'Posts array (required for non-draft)',
                },
                {
                    displayName: 'Start Date (UTC)',
                    name: 'startDate',
                    type: 'dateTime',
                    displayOptions: {
                        show: {
                            operation: ['getPosts'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'Start date for filtering posts (UTC)',
                },
                {
                    displayName: 'End Date (UTC)',
                    name: 'endDate',
                    type: 'dateTime',
                    displayOptions: {
                        show: {
                            operation: ['getPosts'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'End date for filtering posts (UTC)',
                },
                {
                    displayName: 'Customer',
                    name: 'customer',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['getPosts'],
                        },
                    },
                    default: '',
                    description: 'Customer ID for filtering posts (optional)',
                },
                {
                    displayName: 'Binary Property',
                    name: 'binaryProperty',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['uploadFile'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'Name of the binary property that contains the file data',
                },
                {
                    displayName: 'Post ID',
                    name: 'postId',
                    type: 'string',
                    displayOptions: {
                        show: {
                            operation: ['deletePost'],
                        },
                    },
                    default: '',
                    required: true,
                    description: 'ID of the post to delete',
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        const length = items.length;
        let responseData = undefined;
        for (let i = 0; i < length; i++) {
            try {
                const operation = this.getNodeParameter('operation', i);
                if (operation === 'createPost') {
                    const type = this.getNodeParameter('type', i);
                    const shortLink = this.getNodeParameter('shortLink', i);
                    const date = this.getNodeParameter('date', i);
                    const tagsParam = this.getNodeParameter('tags', i, {});
                    const postsParam = this.getNodeParameter('posts', i, {});
                    const tags = tagsParam.tag
                        ? tagsParam.tag.map((tag) => ({
                            value: tag.value,
                            label: tag.label,
                        }))
                        : [];
                    const posts = postsParam.post
                        ? postsParam.post.map((post) => {
                            var _a, _b;
                            const settings = {};
                            if (((_a = post.settings) === null || _a === void 0 ? void 0 : _a.setting) && post.settings.setting.length > 0) {
                                post.settings.setting.forEach((setting) => {
                                    let value;
                                    switch (setting.valueType) {
                                        case 'string':
                                            value = setting.stringValue;
                                            break;
                                        case 'number':
                                            value = setting.numberValue;
                                            break;
                                        case 'boolean':
                                            value = setting.booleanValue;
                                            break;
                                        case 'json':
                                            try {
                                                value = JSON.parse(setting.jsonValue);
                                            }
                                            catch {
                                                value = setting.jsonValue;
                                            }
                                            break;
                                        default:
                                            value = setting.stringValue;
                                    }
                                    settings[setting.key] = value;
                                });
                            }
                            const value = ((_b = post.value) === null || _b === void 0 ? void 0 : _b.contentItem)
                                ? post.value.contentItem.map((item) => {
                                    var _a;
                                    return ({
                                        content: item.content,
                                        id: item.id || '',
                                        image: ((_a = item.image) === null || _a === void 0 ? void 0 : _a.imageItem)
                                            ? item.image.imageItem.map((img) => ({
                                                id: img.id,
                                                path: img.path,
                                            }))
                                            : [],
                                    });
                                })
                                : [];
                            return {
                                integration: {
                                    id: post.integrationId,
                                },
                                value,
                                group: post.group || '',
                                settings,
                            };
                        })
                        : [];
                    const body = {
                        type,
                        shortLink,
                        date,
                        tags,
                        posts,
                    };
                    responseData = await GenericFunctions_1.postizApiRequest.call(this, 'POST', '/posts', body);
                }
                if (operation === 'getPosts') {
                    const startDate = this.getNodeParameter('startDate', i);
                    const endDate = this.getNodeParameter('endDate', i);
                    const customer = this.getNodeParameter('customer', i);
                    const query = {
                        startDate,
                        endDate,
                        ...(customer && { customer }),
                    };
                    responseData = await GenericFunctions_1.postizApiRequest.call(this, 'GET', '/posts', {}, query);
                }
                if (operation === 'uploadFile') {
                    const binaryPropertyName = this.getNodeParameter('binaryProperty', i);
                    const binaryData = this.helpers.assertBinaryData(i, binaryPropertyName);
                    const dataBuffer = await this.helpers.getBinaryDataBuffer(i, binaryPropertyName);
                    const mimeType = binaryData.mimeType || 'application/octet-stream';
                    if (!dataBuffer || !binaryData) {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Item is not of type "binary" or does not contain the expected properties: data, mimeType, fileName`, { itemIndex: i });
                    }
                    const blob = new Blob([dataBuffer], {
                        type: mimeType,
                    });
                    const formData = new FormData();
                    formData.append('file', blob, binaryData.fileName);
                    responseData = await GenericFunctions_1.postizApiRequest.call(this, 'POST', '/upload', formData);
                }
                if (operation === 'getIntegrations') {
                    responseData = await GenericFunctions_1.postizApiRequest.call(this, 'GET', '/integrations');
                }
                if (operation === 'deletePost') {
                    const postId = this.getNodeParameter('postId', i);
                    responseData = await GenericFunctions_1.postizApiRequest.call(this, 'DELETE', `/posts/${postId}`);
                }
                if (operation === 'generateVideo') {
                    const videoType = this.getNodeParameter('videoType', i);
                    const output = this.getNodeParameter('output', i);
                    const customParametersParam = this.getNodeParameter('customParameters', i, {});
                    const body = {
                        type: videoType,
                        output,
                        customParams: {},
                    };
                    if (customParametersParam.parameter && customParametersParam.parameter.length > 0) {
                        customParametersParam.parameter.forEach((param) => {
                            try {
                                body.customParams[param.key] = JSON.parse(param.value);
                            }
                            catch {
                                body.customParams[param.key] = param.value;
                            }
                        });
                    }
                    responseData = await GenericFunctions_1.postizApiRequest.call(this, 'POST', '/generate-video', body);
                }
                if (operation === 'videoFunction') {
                    const functionName = this.getNodeParameter('functionName', i);
                    const identifier = this.getNodeParameter('identifier', i);
                    const additionalParametersParam = this.getNodeParameter('additionalParameters', i, {});
                    const body = {
                        functionName,
                        identifier,
                    };
                    if (additionalParametersParam.parameter &&
                        additionalParametersParam.parameter.length > 0) {
                        additionalParametersParam.parameter.forEach((param) => {
                            body[param.key] = param.value;
                        });
                    }
                    responseData = await GenericFunctions_1.postizApiRequest.call(this, 'POST', '/video/function', body);
                }
                const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                returnData.push(...executionData);
            }
            catch (error) {
                if (this.continueOnFail()) {
                    const executionErrorData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: (error === null || error === void 0 ? void 0 : error.description) || (error === null || error === void 0 ? void 0 : error.message) || error }), { itemData: { item: i } });
                    returnData.push(...executionErrorData);
                    continue;
                }
                throw error;
            }
        }
        return [returnData];
    }
}
exports.Postiz = Postiz;
