// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/getCrawlStatus/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = exports.displayName = exports.name = void 0;
const common_1 = require("../common");
exports.name = 'getCrawlStatus';
exports.displayName = 'Get crawl status';
exports.operationName = 'getCrawlStatus';
exports.resourceName = 'Crawling';
function createCrawlIdProperty() {
    return {
        displayName: 'Crawl ID',
        name: 'crawlId',
        type: 'string',
        required: true,
        default: '',
        description: 'ID of the crawl job to get status for',
        placeholder: '1234abcd-5678-efgh-9012-ijklmnopqrst',
        routing: {
            request: {
                url: '=/crawl/{{$value}}',
            },
        },
        displayOptions: {
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createGetCrawlStatusProperties() {
    return [
        (0, common_1.createOperationNotice)(exports.resourceName, exports.name, 'GET'),
        createCrawlIdProperty(),
    ];
}
const { options, properties } = (0, common_1.buildApiProperties)(exports.name, exports.displayName, createGetCrawlStatusProperties());
exports.options = options;
exports.properties = properties;
options.routing = {
    request: {
        method: 'GET',
    },
    output: {
        postReceive: [
            {
                type: 'setKeyValue',
                properties: {
                    data: '={{$response.body}}',
                },
            },
        ],
    },
};
