// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/Firecrawl.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Firecrawl = void 0;
const methods_1 = require("./methods");
const properties_1 = require("./properties");
class Firecrawl {
    constructor() {
        this.description = {
            displayName: 'Firecrawl',
            name: 'firecrawl',
            icon: 'file:firecrawl.svg',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Scrape, crawl, map, search, and extract structured data from websites using Firecrawl API',
            defaults: {
                name: 'Firecrawl',
            },
            usableAsTool: true,
            inputs: `={{["main"]}}`,
            outputs: `={{["main"]}}`,
            credentials: [
                {
                    name: 'firecrawlApi',
                    required: true,
                },
            ],
            requestDefaults: {
                baseURL: '={{ $credentials.baseUrl || "https://api.firecrawl.dev/v2" }}',
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                },
                body: {
                    integration: 'n8n',
                },
            },
            properties: properties_1.allProperties,
        };
        this.methods = methods_1.allMethods;
    }
}
exports.Firecrawl = Firecrawl;
