// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/helpers/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.postReceiveActionBinaryData = exports.preSendActionCustomBody = exports.aggregateNodeMethods = exports.buildPropertiesWithMethods = exports.buildPropertiesWithOptions = exports.processBinaryResponseData = exports.isBinaryResponse = exports.getFileExtensionFromContentType = exports.getFileTypeFromContentType = exports.getResponseContentType = exports.mergeCustomBodyWithRequest = void 0;
__exportStar(require("./types"), exports);
const requestModifiers_1 = require("./http/requestModifiers");
Object.defineProperty(exports, "mergeCustomBodyWithRequest", { enumerable: true, get: function () { return requestModifiers_1.mergeCustomBodyWithRequest; } });
const responseHandlers_1 = require("./http/responseHandlers");
Object.defineProperty(exports, "getResponseContentType", { enumerable: true, get: function () { return responseHandlers_1.getResponseContentType; } });
Object.defineProperty(exports, "getFileTypeFromContentType", { enumerable: true, get: function () { return responseHandlers_1.getFileTypeFromContentType; } });
Object.defineProperty(exports, "getFileExtensionFromContentType", { enumerable: true, get: function () { return responseHandlers_1.getFileExtensionFromContentType; } });
Object.defineProperty(exports, "isBinaryResponse", { enumerable: true, get: function () { return responseHandlers_1.isBinaryResponse; } });
Object.defineProperty(exports, "processBinaryResponseData", { enumerable: true, get: function () { return responseHandlers_1.processBinaryResponseData; } });
const propertyBuilders_1 = require("./utils/propertyBuilders");
Object.defineProperty(exports, "buildPropertiesWithOptions", { enumerable: true, get: function () { return propertyBuilders_1.buildPropertiesWithOptions; } });
Object.defineProperty(exports, "buildPropertiesWithMethods", { enumerable: true, get: function () { return propertyBuilders_1.buildPropertiesWithMethods; } });
Object.defineProperty(exports, "aggregateNodeMethods", { enumerable: true, get: function () { return propertyBuilders_1.aggregateNodeMethods; } });
exports.preSendActionCustomBody = requestModifiers_1.mergeCustomBodyWithRequest;
exports.postReceiveActionBinaryData = responseHandlers_1.processBinaryResponseData;
