// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/getExtractStatus/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = void 0;
const common_1 = require("../common");
const name = 'getExtractStatus';
const displayName = 'Get Extract Status';
exports.operationName = 'getExtractStatus';
exports.resourceName = 'Extract';
function createExtractIdProperty() {
    return {
        displayName: 'Extract ID',
        name: 'extractId',
        type: 'string',
        required: true,
        default: '',
        description: 'ID of the extract job to get status for',
        placeholder: '1234abcd-5678-efgh-9012-ijklmnopqrst',
        routing: {
            request: {
                url: '=/extract/{{$value}}',
            },
        },
        displayOptions: {
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createGetExtractStatusProperties() {
    return [(0, common_1.createOperationNotice)(exports.resourceName, name, 'GET'), createExtractIdProperty()];
}
const { options, properties } = (0, common_1.buildApiProperties)(name, displayName, createGetExtractStatusProperties());
exports.options = options;
exports.properties = properties;
options.routing = {
    request: {
        method: 'GET',
    },
    output: {
        postReceive: [
            {
                type: 'setKeyValue',
                properties: {
                    data: '={{$response.body}}',
                },
            },
        ],
    },
};
