// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/getCrawlErrors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.displayName = exports.name = void 0;
const common_1 = require("../common");
exports.name = 'getCrawlErrors';
exports.displayName = 'Get crawl errors';
exports.resourceName = 'Crawling';
function createCrawlIdProperty() {
    return {
        displayName: 'Crawl ID',
        name: 'crawlId',
        type: 'string',
        required: true,
        default: '',
        description: 'ID of the crawl job',
        routing: {
            request: {
                url: '=/crawl/{{$value}}/errors',
            },
        },
        displayOptions: {
            show: {
                resource: [exports.resourceName],
                operation: [exports.name],
            },
        },
    };
}
function createProperties() {
    return [(0, common_1.createOperationNotice)(exports.resourceName, exports.name, 'GET'), createCrawlIdProperty()];
}
const { options, properties } = (0, common_1.buildApiProperties)(exports.name, exports.displayName, createProperties());
exports.options = options;
exports.properties = properties;
options.routing = {
    request: {
        method: 'GET',
    },
};
