// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/extract/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = void 0;
const common_1 = require("../common");
const name = 'extract';
const displayName = 'Extract structured data from websites using AI';
exports.operationName = 'extract';
exports.resourceName = 'Extract';
function createUrlsProperty() {
    return {
        displayName: 'URLs',
        name: 'urls',
        type: 'fixedCollection',
        default: {},
        typeOptions: {
            multipleValues: true,
        },
        description: 'URLs to extract data from. Supports glob patterns like "https://example.com/products/*" to extract from multiple pages matching the pattern.',
        placeholder: 'Add URL pattern',
        options: [
            {
                displayName: 'Items',
                name: 'items',
                values: [
                    {
                        displayName: 'URL',
                        name: 'url',
                        type: 'string',
                        default: '',
                        placeholder: 'https://example.com/*',
                        description: 'URL to extract data from (supports glob format)',
                    },
                ],
            },
        ],
        routing: {
            request: {
                body: {
                    urls: '={{$value.items ? $value.items.map(item => item.url) : []}}',
                },
            },
        },
        displayOptions: {
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createPromptProperty() {
    return {
        displayName: 'Prompt',
        name: 'prompt',
        type: 'string',
        default: '',
        description: 'Natural language instructions for the AI to guide data extraction. Be specific about what data you want (e.g., "Extract product names, prices, and descriptions from this e-commerce page").',
        routing: {
            request: {
                body: {
                    prompt: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createSchemaProperty() {
    return {
        displayName: 'Schema',
        name: 'schema',
        type: 'json',
        default: '{\n  "property1": "string",\n  "property2": "number"\n}',
        description: 'JSON Schema defining the structure of extracted data. Specify property names and types (string, number, boolean, array, object). The AI will extract data matching this schema.',
        routing: {
            request: {
                body: {
                    schema: '={{ JSON.parse($value) }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createEnableWebSearchProperty() {
    return {
        displayName: 'Enable Web Search',
        name: 'enableWebSearch',
        type: 'boolean',
        default: false,
        description: 'Whether to allow the AI to search the web for additional information not found on the page. Useful for enriching extracted data with external context.',
        routing: {
            request: {
                body: {
                    enableWebSearch: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createIgnoreSitemapProperty() {
    return {
        displayName: 'Ignore Sitemap',
        name: 'ignoreSitemap',
        type: 'boolean',
        default: true,
        description: 'Whether to skip the sitemap when discovering URLs matching your glob pattern. Enable for faster extraction when you know the exact URL patterns you want.',
        routing: {
            request: {
                body: {
                    ignoreSitemap: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createIncludeSubdomainsProperty() {
    return {
        displayName: 'Include Subdomains',
        name: 'includeSubdomains',
        type: 'boolean',
        default: false,
        description: 'Whether to extract data from subdomains matching your URL patterns. Enable to include blog.example.com or docs.example.com when extracting from example.com.',
        routing: {
            request: {
                body: {
                    includeSubdomains: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createShowSourcesProperty() {
    return {
        displayName: 'Show Sources',
        name: 'showSources',
        type: 'boolean',
        default: false,
        description: 'Whether to include source URLs and page sections where each piece of data was found. Useful for verification, citation, or debugging extraction results.',
        routing: {
            request: {
                body: {
                    showSources: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            hide: {
                useCustomBody: [true],
            },
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createAdditionalFieldsProperty(operation) {
    return {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        description: 'Additional fields to send in the request body',
        options: [
            {
                displayName: 'Custom Properties (JSON)',
                name: 'customProperties',
                type: 'json',
                default: '{}',
                description: 'Custom JSON properties to add to the request body',
            },
        ],
        routing: {
            request: {
                body: {
                    additionalFields: '={{ $value }}',
                },
            },
            send: {
                preSend: [
                    async function (requestOptions) {
                        if (typeof requestOptions.body !== 'object' || !requestOptions.body) {
                            return requestOptions;
                        }
                        const body = requestOptions.body;
                        const additionalFields = body.additionalFields;
                        if (additionalFields) {
                            if (additionalFields.customProperties) {
                                try {
                                    const customProps = JSON.parse(additionalFields.customProperties);
                                    Object.assign(requestOptions.body, customProps);
                                }
                                catch (error) {
                                }
                            }
                            delete body.additionalFields;
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        displayOptions: {
            show: {
                operation: [operation],
                useCustomBody: [true],
            },
        },
    };
}
function createExtractProperties() {
    return [
        (0, common_1.createOperationNotice)(exports.resourceName, name, 'POST'),
        createUrlsProperty(),
        createPromptProperty(),
        createSchemaProperty(),
        createIgnoreSitemapProperty(),
        createIncludeSubdomainsProperty(),
        createEnableWebSearchProperty(),
        createShowSourcesProperty(),
        (0, common_1.createScrapeOptionsProperty)(exports.operationName, true, false, exports.resourceName),
    ];
}
const { options, properties } = (0, common_1.buildApiProperties)(name, displayName, createExtractProperties());
exports.options = options;
exports.properties = properties;
properties.push(createAdditionalFieldsProperty(name));
