// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/getAgentStatus/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.operationName = void 0;
const common_1 = require("../common");
const name = 'getAgentStatus';
const displayName = 'Get Agent Status';
exports.operationName = 'getAgentStatus';
exports.resourceName = 'Agent';
function createAgentIdProperty() {
    return {
        displayName: 'Agent Job ID',
        name: 'agentId',
        type: 'string',
        required: true,
        default: '',
        description: 'ID of the agent job to get status for. Returns status (processing, completed, failed), extracted data, credits used, and expiration time. Job results are available for 24 hours after completion.',
        placeholder: '1234abcd-5678-efgh-9012-ijklmnopqrst',
        routing: {
            request: {
                url: '=/agent/{{$value}}',
            },
        },
        displayOptions: {
            show: {
                resource: [exports.resourceName],
                operation: [exports.operationName],
            },
        },
    };
}
function createGetAgentStatusProperties() {
    return [(0, common_1.createOperationNotice)(exports.resourceName, name, 'GET'), createAgentIdProperty()];
}
const { options, properties } = (0, common_1.buildApiProperties)(name, displayName, createGetAgentStatusProperties());
exports.options = options;
exports.properties = properties;
options.routing = {
    request: {
        method: 'GET',
    },
    output: {
        postReceive: [
            {
                type: 'setKeyValue',
                properties: {
                    data: '={{$response.body}}',
                },
            },
        ],
    },
};
