// @mendable/n8n-nodes-firecrawl/dist/nodes/Firecrawl/api/crawlParamsPreview/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.properties = exports.options = exports.resourceName = exports.displayName = exports.name = void 0;
const common_1 = require("../common");
exports.name = 'crawlParamsPreview';
exports.displayName = 'Preview crawl params from prompt';
exports.resourceName = 'Crawling';
function createPromptProperty() {
    return {
        displayName: 'Prompt',
        name: 'prompt',
        type: 'string',
        default: '',
        required: true,
        description: 'Natural language prompt describing crawl behavior',
        routing: {
            request: {
                body: {
                    prompt: '={{ $value }}',
                },
            },
        },
        displayOptions: {
            show: {
                resource: [exports.resourceName],
                operation: [exports.name],
            },
        },
    };
}
function createProperties() {
    return [(0, common_1.createOperationNotice)(exports.resourceName, exports.name), (0, common_1.createUrlProperty)(exports.name, undefined, exports.resourceName), createPromptProperty()];
}
const { options, properties } = (0, common_1.buildApiProperties)(exports.name, exports.displayName, createProperties());
exports.options = options;
exports.properties = properties;
options.routing = {
    request: {
        url: '=/crawl/params-preview',
    },
};
