// @mendable/n8n-nodes-firecrawl/dist/credentials/FirecrawlApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirecrawlApi = void 0;
class FirecrawlApi {
    constructor() {
        this.name = 'firecrawlApi';
        this.displayName = 'Firecrawl API';
        this.documentationUrl = 'https://docs.firecrawl.dev/introduction#api-key';
        this.properties = [
            {
                displayName: 'Base URL',
                name: 'baseUrl',
                type: 'string',
                default: 'https://api.firecrawl.dev/v2',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.baseUrl}}',
                url: '/team/credit-usage',
                method: 'GET',
            },
        };
    }
}
exports.FirecrawlApi = FirecrawlApi;
